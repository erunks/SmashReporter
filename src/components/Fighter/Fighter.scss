@import '../../styles/base.scss';

.fighter {
  display: flex;
  position: relative;
}

.fighterCard {  
  background-repeat: no-repeat;
  background-size: cover;
  border-right: 2px solid;
  height: 55px;
  max-width: 100px;
  position: relative;
  width: 90px;

  &:hover {
    z-index: 2;

    .selection {
      @include visible(inline);
    }
  }

  .selection {
    &.player-one {
      @include blink-animation(primary-blink);
      border-color: $primary-selection-color;

      .bottomRight,
      .topLeft {
        @include blink-animation(primary-corner-blink);
      }

      .bottomRight {
        border-top-color: $primary-selection-color;
      }

      .topLeft {
        border-bottom-color: $primary-selection-color;
      }
    }
    &.player-two {
      @include blink-animation(secondary-blink);
      border-color: $secondary-selection-color;

      .bottomRight,
      .topLeft {
        @include blink-animation(secondary-corner-blink);
      }

      .bottomRight {
        border-top-color: $secondary-selection-color;
      }

      .topLeft {
        border-bottom-color: $secondary-selection-color;
      }
    }

    @include absolute-position;
    @include hidden;

    border: 4px solid;
    height: 100%;
    width: 100%;

    .bottomRight,
    .topLeft {
      height: 0;
      position: absolute;
      width: 0;
    }

    .bottomRight {
      border-right: 12px solid transparent;
      border-top: 12px solid;
      left: 0;
      top: 0;
    }

    .topLeft {
      border-bottom: 12px solid;
      border-left: 12px solid transparent;
      bottom: 0;
      right: 0;
    }
  }

  .namePlate {
    background-color: black;
    height: 30px;
    max-width: 100%;
  }

  .name {
    color: white;
    font-size: 12px;
    padding: 0.25em 2px;
    margin: 0;
    overflow: hidden;
    text-align: center;
    text-overflow: ellipsis;
    white-space: nowrap;
    width: 100%;
  }
}

@keyframes primary-blink {
	0% {
    border-color: $primary-selection-color;
  }
  
  50% {
    border-color: white;
  }

  100% {
    border-color: $primary-selection-color;
  }
}

@keyframes primary-corner-blink {
	0% {
    border-bottom-color: $primary-selection-color;
    border-top-color: $primary-selection-color;
  }
  
  50% {
    border-bottom-color: white;
    border-top-color: white;
  }

  100% {
    border-bottom-color: $primary-selection-color;
    border-top-color: $primary-selection-color;
  }
}

@keyframes secondary-blink {
	0% {
    border-color: $secondary-selection-color;
  }
  
  50% {
    border-color: white;
  }

  100% {
    border-color: $secondary-selection-color;
  }
}

@keyframes secondary-corner-blink {
	0% {
    border-bottom-color: $secondary-selection-color;
    border-top-color: $secondary-selection-color;
  }
  
  50% {
    border-bottom-color: white;
    border-top-color: white;
  }

  100% {
    border-bottom-color: $secondary-selection-color;
    border-top-color: $secondary-selection-color;
  }
}
