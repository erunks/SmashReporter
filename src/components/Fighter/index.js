import React from 'react';
import propTypes from 'prop-types';
import { useRecoilState, useRecoilValue } from 'recoil';
import { selectedFightersState } from 'recoils/fighter';
import { selectingPlayerState } from 'recoils/player';
import { buttonize } from 'helpers/jsx-a11y-helper'
import { IKImage } from 'imagekitio-react';
import { URLS } from 'lib/imageAssets';

import './Fighter.scss';

const Fighter = (props) => {
  const [selectedFighters, setSelectedFighters] = useRecoilState(selectedFightersState);
  const selectingPlayer = useRecoilValue(selectingPlayerState);

  const fighterAssetName = (name) => {
    let assetName = name.replace(/[- ]/g, '_');
    assetName = assetName.replace(/&/g, 'and');
    assetName = assetName.replace(/\./g, '');
    return assetName.toLowerCase();
  };

  const setFighter = () => {
    const newSelectedFighters = [...selectedFighters];
    newSelectedFighters[selectingPlayer] = asset_name;
    setSelectedFighters(newSelectedFighters);
  };

  const asset_name = props.asset_name || fighterAssetName(props.name);
  const fighter_thumbnail_icon = props.asset_url || URLS.fighterThumbnailIcon(asset_name);

  return (
    <div className="fighter" { ...buttonize(() => setFighter()) }>
      <div className="fighterCard" alt={`fighter: ${props.name}`}>
        <IKImage
          src={fighter_thumbnail_icon}
          alt={`fighter: ${props.name}`}
          loading="lazy"
          lqip={{ active: true }}
        />
        <div className={`selection ${selectingPlayer === 0 ? 'player-one' : 'player-two'}`}>
          <div className="topLeft"></div>
          <div className="bottomRight"></div>
        </div>
      </div>
    </div>
  );
};

const {
  bool,
  number,
  oneOfType,
  string
} = propTypes;

Fighter.propTypes = {
  dlc: oneOfType([
    bool,
    number
  ]).isRequired,
  id: number.isRequired,
  name: string.isRequired
};

export default Fighter;
