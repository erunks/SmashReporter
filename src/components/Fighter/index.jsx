import React from 'react';
import propTypes from 'prop-types';
import { buttonize } from '../../helpers/jsx-a11y-helper'
import { URLS } from '../../lib/imageAssets';

import './Fighter.scss';

const Fighter = (props) => {
  const fighterAssetName = (name) => {
    let assetName = name.replace(/[- ]/g, '_');
    assetName = assetName.replace(/&/g, 'and');
    assetName = assetName.replace(/\./g, '');
    return assetName.toLowerCase();
  };

  const asset_name = props.asset_name || fighterAssetName(props.name);
  const fighter_thumbnail_icon = props.asset_url || URLS.fighterThumbnailIcon(asset_name);
  const setFighter = () => {
    props.setPlayerFighter(asset_name);
  };

  return (
    <div className="fighter" { ...buttonize(() => setFighter()) }>
      <div className="fighterCard" style={{ backgroundImage: `url(${fighter_thumbnail_icon})` }} alt={`fighter: ${props.name}`}>
        <div className={`selection ${props.selectingPlayer === 0 ? 'player-one' : 'player-two'}`}>
          <div className="topLeft"></div>
          <div className="bottomRight"></div>
        </div>
      </div>
    </div>
  );
};

const {
  bool,
  func,
  number,
  oneOfType,
  string
} = propTypes;

Fighter.propTypes = {
  dlc: oneOfType([
    bool,
    number
  ]).isRequired,
  id: number.isRequired,
  name: string.isRequired,
  selectingPlayer: number.isRequired,
  setPlayerFighter: func.isRequired
};

export default Fighter;
