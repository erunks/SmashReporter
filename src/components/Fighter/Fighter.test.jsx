import React from 'react';
import { shallow } from 'enzyme';
import { spy } from 'sinon';
import Fighter from './index.jsx';

describe('<Fighter>', () => {
  const props = {
    dlc: 0,
    id: 1,
    name: 'Mario',
    setPlayerFighter: spy()
  };
  const container = shallow(<Fighter selectingPlayer={0} {...props}/>);

  it('should match the snapshot', () => {
    expect(container.html()).toMatchSnapshot();
  });

  describe('when the fighter is clicked', () => {
    it('should call setPlayerFighter with the asset_name', () => {
      const asset_name = props.name.toLowerCase();
      const fighter = container.find('.fighter');
      fighter.simulate('click');

      expect(props.setPlayerFighter.calledWith(asset_name)).toBe(true);
    });
  });

  describe('when selectingPlayer is player 2', () => {
    const container = shallow(<Fighter selectingPlayer={1} {...props} />);

    it('should set the selection class appropriately', () => {
      const selection = container.find('.selection');

      expect(selection.hasClass('player-two')).toBe(true);
    })
  });
});
