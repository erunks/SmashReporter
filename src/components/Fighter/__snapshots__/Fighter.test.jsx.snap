// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Fighter> should match the snapshot 1`] = `"<div class=\\"fighter\\" role=\\"button\\"><div class=\\"fighterCard\\" style=\\"background-image:url(https://www.smashbros.com/assets_v2/img/fighter/thumb_a/mario.png)\\" alt=\\"fighter: Mario\\"><div class=\\"selection player-one\\"><div class=\\"topLeft\\"></div><div class=\\"bottomRight\\"></div></div></div></div>"`;
